src/main/java/net/neoforged/neoforge/capabilities/BlockCapability.java

```diff
@@ -0,0 +1,133 @@
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A {@code BlockCapability} gives flexible access to objects of type {@code T} located in the world.
+ *
+ * <h3>Querying a block capability</h3>
+ * <p>
+ * To get an object of type {@code T}, use {@link Level#getCapability(BlockCapability, BlockPos, Object)}.
+ * For example, to query an item handler in the world, from a specific side:
+ * 
+ * <pre>{@code
+ * Level level = ...;
+ * BlockPos pos = ...;
+ * Direction side = ...;
+ *
+ * IItemHandler maybeHandler = level.getCapability(ForgeCapabilities.ItemHandler.BLOCK, pos, side);
+ * if (maybeHandler != null) {
+ *     // Use maybeHandler
+ * }
+ * }</pre>
+ * 
+ * <p>
+ * For repeated queries at a specific position, use {@link BlockCapabilityCache} to improve performance.
+ *
+ * <h3>Providing a block capability</h3>
+ * <p>
+ * To provide objects of type {@code T}, register providers to {@link RegisterCapabilitiesEvent}. For example:
+ * 
+ * <pre>{@code
+ * modBus.addListener((RegisterCapabilitiesEvent event) -> {
+ *     event.registerBlockEntity(
+ *         ForgeCapabilities.ITEM_HANDLER_BLOCK, // capability to register for
+ *         MY_BLOCK_ENTITY_TYPE,
+ *         (myBlockEntity, side) -> <return the IItemHandler for myBlockEntity and side>);
+ * });
+ * }</pre>
+ * 
+ * <p>
+ * <b>If a previously returned capability is not valid anymore, or if a new capability is available,
+ * {@link Level#invalidateCapabilities(BlockPos)} MUST be called to notify the caches (see below).</b>
+ *
+ * @param <T> Type of queried objects.
+ * @param <C> Type of the additional context.
+ */
+public final class BlockCapability<T, C> {
+    /**
+     * Creates a new block capability, or get it if it already exists.
+     *
+     * @param name         Name of the capability
+     * @param typeClass    Type of the queried API
+     * @param contextClass Type of the additional context
+     */
+    public static <T, C> BlockCapability<T, C> create(ResourceLocation name, Class<T> typeClass, Class<C> contextClass) {
+        return (BlockCapability<T, C>) registry.create(name, typeClass, contextClass);
+    }
+
+    /**
+     * Creates a new block capability with {@code Void} context, or get it if it already exists.
+     * This should be used for capabilities that do not require any additional context.
+     *
+     * @see #create(ResourceLocation, Class, Class)
+     */
+    public static <T> BlockCapability<T, Void> createVoid(ResourceLocation name, Class<T> typeClass) {
+        return create(name, typeClass, void.class);
+    }
+
+    public ResourceLocation name() {
+        return name;
+    }
+
+    public Class<T> typeClass() {
+        return typeClass;
+    }
+
+    public Class<C> contextClass() {
+        return contextClass;
+    }
+
+    // INTERNAL
+    private static final CapabilityRegistry<BlockCapability<?, ?>> registry = new CapabilityRegistry<>(BlockCapability::new);
```

Comment:
@Technici4n would you consider making a read-only accessor on this?
You know how I said loop over all blocks and add all capabilities and lazy check on them.
Having access to this (without reflection/mixins/AT) would be nice. I wouldn't say make it modifiable, but a read-only iterator?

Reply:
yes, @covers1624 also requested something similar

Reply:
added `getAll` static methods ;)